democracy/src/lib.rs

```diff
@@ -0,0 +1,427 @@
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight({10000})]
+		pub fn submit_proposal(
+			origin: OriginFor<T>,
+			proposal_action: ProposalAction,
+		) -> DispatchResultWithPostInfo {
+			if Self::enactment_queue(proposal_action.get_identifier()).is_some() {
+				return Err(Error::<T>::ProposalWaitingForEnactment.into())
+			}
+			let _sender = ensure_signed(origin)?;
+			let cindex = <encointer_scheduler::Pallet<T>>::current_ceremony_index();
+			let current_proposal_id = Self::proposal_count();
+			let next_proposal_id = current_proposal_id
+				.checked_add(1u128)
+				.ok_or(Error::<T>::ProposalIdOutOfBounds)?;
+			let current_block = frame_system::Pallet::<T>::block_number();
+			let proposal = Proposal {
+				start: current_block,
+				start_cindex: cindex,
+				state: ProposalState::Ongoing,
+				action: proposal_action,
+			};
+			<Proposals<T>>::insert(next_proposal_id, proposal);
+			<ProposalCount<T>>::put(next_proposal_id);
+			<Tallies<T>>::insert(next_proposal_id, Tally { turnout: 0, ayes: 0 });
+			Ok(().into())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight({10000})]
+		pub fn vote(
+			origin: OriginFor<T>,
+			proposal_id: ProposalIdType,
+			vote: Vote,
+			reputations: ReputationVecOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			let tally = <Tallies<T>>::get(proposal_id).ok_or(Error::<T>::InexistentProposal)?;
+			let valid_reputations = Self::valid_reputations(proposal_id, &sender, &reputations)?;
+			let num_votes = valid_reputations.len() as u128;
+
+			let ayes = match vote {
+				Vote::Aye => num_votes,
+				Vote::Nay => 0,
+			};
+
+			let new_tally = Tally {
+				turnout: tally
+					.turnout
+					.checked_add(num_votes)
+					.ok_or(Error::<T>::VoteCountOverflow)?,
+				ayes: tally.ayes.checked_add(ayes).ok_or(Error::<T>::VoteCountOverflow)?,
+			};
+
+			<Tallies<T>>::insert(proposal_id, new_tally);
+			for community_ceremony in valid_reputations {
+				<VoteEntries<T>>::insert(proposal_id, (&sender, community_ceremony), ());
+			}
+
+			Ok(().into())
+		}
+	}
+	impl<T: Config> Pallet<T> {
+		fn relevant_cindexes(
+			proposal_id: ProposalIdType,
+		) -> Result<Vec<CeremonyIndexType>, Error<T>> {
+			let reputation_lifetime = <encointer_ceremonies::Pallet<T>>::reputation_lifetime();
+			let proposal = Self::proposals(proposal_id).ok_or(Error::<T>::InexistentProposal)?;
+			Ok(((proposal
+				.start_cindex
+				.saturating_sub(reputation_lifetime)
+				.saturating_add(T::ProposalLifetimeCycles::get()))..=
+				(proposal.start_cindex.saturating_sub(2)))
+				.collect::<Vec<CeremonyIndexType>>())
+		}
+		/// Returns the reputations that
+		/// 1. are valid
+		/// 2. have not been used to vote for proposal_id
+		/// 3. originate in the correct community (for Community AccessPolicy)
+		/// 4. are within proposal.start_cindex - reputation_lifetime + proposal_lifetime and proposal.start_cindex - 2
+		pub fn valid_reputations(
```

Comment:
reading your fn doc above, I could imagine that ` eligible_reputations`
would be a more precise naming.

Reply:
question: If ppl change their minds, this logic will prevent changing the vote of already committed reputation, right?

Reply:
renamed. yes no changing minds in the current protocol

Reply:
that's fine. can be declared as such. and I'm not even sure we want to change that